dataRetriever/dataPool/proofsCache/proofsCache.go

```diff
@@ -43,39 +45,66 @@ func (pc *proofsCache) addProof(proof data.HeaderProofHandler) {
 		return
 	}
 
-	pc.mutProofsCache.Lock()
-	defer pc.mutProofsCache.Unlock()
+	pc.insertProofByNonce(proof)
 
-	pc.proofsByNonce = append(pc.proofsByNonce, &proofNonceMapping{
-		headerHash: string(proof.GetHeaderHash()),
-		nonce:      proof.GetHeaderNonce(),
-	})
+	pc.mutProofsByHash.Lock()
+	pc.proofsByHash[string(proof.GetHeaderHash())] = proof
+	pc.mutProofsByHash.Unlock()
+}
 
-	sort.Slice(pc.proofsByNonce, func(i, j int) bool {
-		return pc.proofsByNonce[i].nonce < pc.proofsByNonce[j].nonce
-	})
+func (pc *proofsCache) insertProofByNonce(proof data.HeaderProofHandler) {
+	pc.mutProofsByNonce.Lock()
+	defer pc.mutProofsByNonce.Unlock()
 
-	pc.proofsByHash[string(proof.GetHeaderHash())] = proof
+	if len(pc.proofsByNonceBuckets) == 0 {
+		pc.insertInNewBucket(proof)
+		return
+	}
+
+	headBucket := pc.proofsByNonceBuckets[0]
+
+	if headBucket.isFull() {
+		pc.insertInNewBucket(proof)
+		return
+	}
+
+	headBucket.insert(proof)
+}
+
+func (pc *proofsCache) insertInNewBucket(proof data.HeaderProofHandler) {
+	bucket := newProofBucket(pc.bucketSize)
+	bucket.insert(proof)
+
+	pc.proofsByNonceBuckets = append([]*proofNonceBucket{bucket}, pc.proofsByNonceBuckets...)
 }
 
 func (pc *proofsCache) cleanupProofsBehindNonce(nonce uint64) {
 	if nonce == 0 {
 		return
 	}
 
-	pc.mutProofsCache.Lock()
-	defer pc.mutProofsCache.Unlock()
+	pc.mutProofsByNonce.Lock()
```

Comment:
If cleanup procedure does not happen extremely often (~rare event), maybe both locks can happen one near the other - e.g. lock & unlock `mutProofsByHash` for all buckets at once (I think this was Sorin's internal suggestion).